### PR TITLE
[SPARK-29212][ML][PYSPARK] Add common classes without using JVM backend

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

Implement common base ML classes (`Predictor`, `PredictionModel`, `Classifier`, `ClasssificationModel` `ProbabilisticClassifier`, `ProbabilisticClasssificationModel`, `Regressor`, `RegrssionModel`) for non-Java backends.

Note

- `Predictor` and `JavaClassifier` should be abstract as `_fit` method is not implemented.
- `PredictionModel` should be abstract as `_transform` is not implemented.

### Why are the changes needed?

To provide extensions points for non-JVM algorithms, as well as a public (as opposed to `Java*` variants, which are commonly described in docstrings as private) hierarchy which can be used to distinguish between different classes of predictors.

For longer discussion see [SPARK-29212](https://issues.apache.org/jira/browse/SPARK-29212) and / or https://github.com/apache/spark/pull/25776.

### Does this PR introduce any user-facing change?

It adds new base classes as listed above, but effective interfaces (method resolution order notwithstanding) stay the same. 

Additionally "private" `Java*` classes in`ml.regression` and `ml.classification` have been renamed to follow PEP-8 conventions (added leading underscore).

It is for discussion if the same should be done to equivalent classes from `ml.wrapper`.

If we take `JavaClassifier` as an example, type hierarchy will change from 

old pyspark ml classification JavaClassifier

to 

new pyspark ml classification _JavaClassifier

Similarly the old model

old pyspark ml classification JavaClassificationModel

will become 

new pyspark ml classification _JavaClassificationModel

 

### How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by zero323** [comment]
#898de5eb562b359fb23130cf8584b2b836a20d87 

- `Predictor` and `JavaClassifier` should be abstract as `_fit` method is not implemented.
- `PredictionModel` should be abstract as `_transform` is not implemented.

**Comment by zero323** [comment]
CC @huaxingao @srowen Forwarding, since you participated in the discussion.

**Comment by zhengruifeng** [comment]
In general LGTM, pending update version info 3.0.0 to 3.1.0
Also ping @srowen @huaxingao  Could you help double check ?

File: python/pyspark/ml/__init__.py [hunk 1]
```diff
@@ -19,13 +19,16 @@
...
+    Transformer, UnaryTransformer
 from pyspark.ml.pipeline import Pipeline, PipelineModel
 from pyspark.ml import classification, clustering, evaluation, feature, fpm, \
     image, pipeline, recommendation, regression, stat, tuning, util, linalg, param
 
 __all__ = [
-    "Transformer", "UnaryTransformer", "Estimator", "Model", "Pipeline", "PipelineModel",
+    "Transformer", "UnaryTransformer", "Estimator", "Model",
+    "Predictor", "PredictionModel",
+    "Pipeline", "PipelineModel",
```
Comment (huaxingao) [inline python/pyspark/ml/__init__.py]: super nit: combine these two lines?

File: python/pyspark/ml/classification.py [hunk 2]
```diff
@@ -74,52 +78,56 @@ def setRawPredictionCol(self, value):
...
+        raise NotImplementedError()
 
 
-class _JavaProbabilisticClassifierParams(HasProbabilityCol, HasThresholds, _JavaClassifierParams):
+class _ProbabilisticClassifierParams(HasProbabilityCol, HasThresholds, _ClassifierParams):
     """
-    Params for :py:class:`JavaProbabilisticClassifier` and
-    :py:class:`JavaProbabilisticClassificationModel`.
+    Params for :py:class:`_JavaProbabilisticClassifier` and
+    :py:class:`_JavaProbabilisticClassificationModel`.
```
Comment (huaxingao) [inline python/pyspark/ml/classification.py]: should these be ```ProbabilisticClassifier``` and ```ProbabilisticClassificationModel```?

**Comment by huaxingao** [comment]
The changes look good to me.

**Comment by zero323** [comment]
> In general LGTM, pending update version info 3.0.0 to 3.1.0

Yeah, I was thinking about the right way to tackle this. While we introduce new objects in type hierarchy, the API of the existing classes doesn't change, compared to what we get in 3.0. So I felt that keeping `since` intact makes sense. But I am happy to change it, if you think it is the way to go.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
Merged to master, thanks all!

**Comment by zero323** [comment]
Thanks!

**Pull request closed without merging**